synapsebridgehelpers/export_tables.py

```diff
@@ -0,0 +1,342 @@
+                renamed_source_column = next(
+                        filter(lambda c : c["name"] == source_name, source_cols))
+                renamed_target_column = next(
+                        filter(lambda c : c["name"] == target_name, target_cols))
+                target_schema.removeColumn(renamed_target_column)
+                target_schema.addColumn(renamed_source_column)
+    target_schema = syn.store(target_schema)
+    return target_schema
+
+
+def export_tables(syn, table_mapping, target_project=None, update=True,
+                  reference_col="recordId", copy_file_handles=True, **kwargs):
+    """Export records from one Synapse table to another. Or transfer tables
+    to completely new tables in a separate project.
+
+    Parameters
+    ----------
+    syn : synapseclient.Synapse
+    table_mapping : dict, list, or str
+        If exporting records of one or more tables to other, preexisting tables,
+        table_mapping is a dictionary containing Synapse ID key/value mappings
+        from source to target tables. If exporting table records to not yet
+        created tables in a seperate project, table_mapping can be a list or
+        string.
+    target_project : str
+        If exporting table records to not yet created tables in a seperate
+        project, specify the target projects Synapse ID here.
+    update : bool, defaults to True
+        When exporting records of one or more tables to other, preexisting
+        tables, whether to append new records to these tables or completely
+        overwrite the table records.
+    referenceCol : str
+        If `update` is True, use this column as the table index to determine
+        which records are already present in the target table.
+    copy_file_handles : bool, defaults to True
+        Whether to copy the file handles from the source table to the target
+        table. If you are not the creator of these file handles, this must
+        be set to True if you want to a column containing file handles in your
+        target table.
+    **kwargs
+        Additional arguments to pass to synapsebridgehelpers.query_across_tables
+
+    Returns
+    -------
+    """
+    results = {}
+    if isinstance(table_mapping, (list, str)): # export to brand new tables
+        if target_project is None:
+            raise TypeError("If passing a list to table_mapping, "
+                            "target_project must be set.")
+        source_tables = synapsebridgehelpers.query_across_tables(
+                syn, tables=table_mapping, as_data_frame=True, **kwargs)
+        if isinstance(table_mapping, str):
+            source_table_iter = [(table_mapping, source_tables[0])]
+        else:
+            source_table_iter = zip(table_mapping, source_tables)
+        for source_id, source_table in source_table_iter:
+            source_table_info = syn.get(source_id)
+            source_table_cols = list(syn.getTableColumns(source_id))
+            if copy_file_handles:
```

Comment:
Wrap in try/catch without copying?

Reply:
Do we want to force the user to use the original file handles if they own them?

Reply:
note to self: default copy_file_handles to false, copy if there is an exception thrown.

Reply:
I implemented this by defaulting `copy_file_handles` to None. If `copy_file_handles` is None, we will try to export the table, and if that fails we will copy the file handles then attempt export again. If the user specifically passes `copy_file_handles` = False we will raise the exception if there is one during export. If `copy_file_handles` = True we will always copy file handles.